In [1]:
import pandas as pd
import numpy as np
from pathlib import Path

# Set up paths
PROJECT_ROOT = Path.cwd().parent
PROCESSED_DIR = PROJECT_ROOT / "data" / "processed"

# Load cleaned data
cleaned_file = PROCESSED_DIR / "cleaned_cancer_data.csv"
df = pd.read_csv(cleaned_file)

print(f"Loaded cleaned data: {df.shape[0]:,} rows × {df.shape[1]} columns")
print(f"Columns: {list(df.columns)}")
print("\nFirst few rows:")
print(df.head())

Loaded cleaned data: 6 rows × 5 columns
Columns: ['Year', 'Deaths', 'Population', 'Age-Adjusted Rate', 'Crude Rate']

First few rows:
     Year  Deaths  Population  Age-Adjusted Rate  Crude Rate
0  2018.0  1841.0  81956393.0                2.2         2.2
1  2019.0  1704.0  81679568.0                2.1         2.1
2  2020.0  1702.0  82873432.0                2.0         2.1
3  2021.0  1722.0  82201873.0                2.1         2.1
4  2022.0  1772.0  81903660.0                2.1         2.2


In [4]:
# Year-over-year death trends
print("=== Year-over-Year Death Trends ===")
df_sorted = df.sort_values('Year')

for i in range(1, len(df_sorted)):
    prev_year = df_sorted.iloc[i-1]
    curr_year = df_sorted.iloc[i]
    
    deaths_change = curr_year['Deaths'] - prev_year['Deaths']
    deaths_pct_change = (deaths_change / prev_year['Deaths']) * 100
    
    print(f"{prev_year['Year']} → {curr_year['Year']}:")
    print(f"  Deaths: {prev_year['Deaths']:,} → {curr_year['Deaths']:,} ({deaths_pct_change:+.1f}%)")
    print(f"  Change: {deaths_change:+,} deaths")

=== Year-over-Year Death Trends ===
2018.0 → 2019.0:
  Deaths: 1,841.0 → 1,704.0 (-7.4%)
  Change: -137.0 deaths
2019.0 → 2020.0:
  Deaths: 1,704.0 → 1,702.0 (-0.1%)
  Change: -2.0 deaths
2020.0 → 2021.0:
  Deaths: 1,702.0 → 1,722.0 (+1.2%)
  Change: +20.0 deaths
2021.0 → 2022.0:
  Deaths: 1,722.0 → 1,772.0 (+2.9%)
  Change: +50.0 deaths
2022.0 → 2023.0:
  Deaths: 1,772.0 → 1,829.0 (+3.2%)
  Change: +57.0 deaths


## Year-over-Year Mortality Trends (2018–2023)

Between 2018 and 2019, total deaths declined significantly by 7.4% (−137 deaths), representing the largest single-year decrease in the observed period. Mortality remained relatively stable from 2019 to 2020, with only a slight 0.1% decrease.

Beginning in 2020, deaths increased for three consecutive years. From 2020 to 2021, deaths rose by 1.2%, followed by a 2.9% increase from 2021 to 2022, and a further 3.2% increase from 2022 to 2023. By 2023, total deaths reached 1,829 — the highest level in the dataset.

Overall, the data suggest an initial decline in mortality followed by a sustained upward trend from 2020 onward. Further analysis is needed to determine whether this increase reflects population growth, aging demographics, reporting changes, or broader public health factors.

In [3]:
# Population growth analysis
print("=== Population Growth Analysis ===")
df_sorted = df.sort_values('Year')

for i in range(1, len(df_sorted)):
    prev_year = df_sorted.iloc[i-1]
    curr_year = df_sorted.iloc[i]
    
    pop_change = curr_year['Population'] - prev_year['Population']
    pop_pct_change = (pop_change / prev_year['Population']) * 100
    
    print(f"{prev_year['Year']} → {curr_year['Year']}:")
    print(f"  Population: {prev_year['Population']:,} → {curr_year['Population']:,}")
    print(f"  Change: {pop_change:+,} ({pop_pct_change:+.2f}%)")

=== Population Growth Analysis ===
2018.0 → 2019.0:
  Population: 81,956,393.0 → 81,679,568.0
  Change: -276,825.0 (-0.34%)
2019.0 → 2020.0:
  Population: 81,679,568.0 → 82,873,432.0
  Change: +1,193,864.0 (+1.46%)
2020.0 → 2021.0:
  Population: 82,873,432.0 → 82,201,873.0
  Change: -671,559.0 (-0.81%)
2021.0 → 2022.0:
  Population: 82,201,873.0 → 81,903,660.0
  Change: -298,213.0 (-0.36%)
2022.0 → 2023.0:
  Population: 81,903,660.0 → 81,573,961.0
  Change: -329,699.0 (-0.40%)


## Population Trends and Impact on Mortality Patterns (2018–2023)

The population remained relatively stable over the study period, fluctuating within a narrow range of approximately 81.5–82.9 million. While there was a temporary increase in population between 2019 and 2020 (+1.46%), population declined in most other years, including consecutive decreases from 2020 through 2023.

Importantly, the recent rise in total deaths from 2020 onward occurred during a period of flat or declining population. For example, from 2021 to 2023, deaths increased steadily (+1.2%, +2.9%, +3.2%), while the population decreased modestly each year. This divergence suggests that the upward trend in deaths cannot be explained by population growth alone.

Because mortality increased even as the population contracted, the data indicate a genuine increase in mortality burden rather than a change driven by demographic expansion. This reinforces the importance of examining age-adjusted rates to determine whether the observed increase reflects shifts in underlying mortality risk.

In [5]:
# Death rate analysis (deaths per 100,000 population)
print("=== Death Rate Analysis (per 100,000 population) ===")
df_sorted = df.sort_values('Year')

# Calculate death rates per 100,000 population
df_sorted['Death_Rate_per_100k'] = (df_sorted['Deaths'] / df_sorted['Population']) * 100000

print("Death Rates per 100,000 Population:")
for _, row in df_sorted.iterrows():
    print(f"  {row['Year']}: {row['Death_Rate_per_100k']:.2f} per 100,000")

print("\nYear-over-Year Death Rate Changes:")
for i in range(1, len(df_sorted)):
    prev_year = df_sorted.iloc[i-1]
    curr_year = df_sorted.iloc[i]
    
    rate_change = curr_year['Death_Rate_per_100k'] - prev_year['Death_Rate_per_100k']
    rate_pct_change = (rate_change / prev_year['Death_Rate_per_100k']) * 100
    
    print(f"{prev_year['Year']} → {curr_year['Year']}:")
    print(f"  Rate: {prev_year['Death_Rate_per_100k']:.2f} → {curr_year['Death_Rate_per_100k']:.2f} ({rate_pct_change:+.1f}%)")
    print(f"  Change: {rate_change:+.2f} per 100,000")

=== Death Rate Analysis (per 100,000 population) ===
Death Rates per 100,000 Population:
  2018.0: 2.25 per 100,000
  2019.0: 2.09 per 100,000
  2020.0: 2.05 per 100,000
  2021.0: 2.09 per 100,000
  2022.0: 2.16 per 100,000
  2023.0: 2.24 per 100,000

Year-over-Year Death Rate Changes:
2018.0 → 2019.0:
  Rate: 2.25 → 2.09 (-7.1%)
  Change: -0.16 per 100,000
2019.0 → 2020.0:
  Rate: 2.09 → 2.05 (-1.6%)
  Change: -0.03 per 100,000
2020.0 → 2021.0:
  Rate: 2.05 → 2.09 (+2.0%)
  Change: +0.04 per 100,000
2021.0 → 2022.0:
  Rate: 2.09 → 2.16 (+3.3%)
  Change: +0.07 per 100,000
2022.0 → 2023.0:
  Rate: 2.16 → 2.24 (+3.6%)
  Change: +0.08 per 100,000


## Death Rate Analysis: Mortality Burden Independent of Population Size (2018–2023)

Death rates per 100,000 population provide a standardized measure of mortality risk that accounts for population fluctuations. This analysis reveals whether changes in total deaths reflect genuine shifts in mortality burden or simply demographic changes.

### Key Findings:

**Initial Decline (2018–2020):**
- Death rates declined from 2.25 to 2.05 per 100,000 (-9.0% total)
- Largest single-year decrease occurred 2018→2019 (-7.1%)
- Suggests improving mortality outcomes during early period

**Sustained Increase (2020–2023):**
- Death rates rose steadily for three consecutive years
- 2020→2021: +2.0%, 2021→2022: +3.3%, 2022→2023: +3.6%
- By 2023, death rates (2.24) nearly returned to 2018 peak (2.25)

### Critical Insight:

The upward mortality trend from 2020 onward is **more pronounced** when examining death rates rather than raw death counts. Death rates increased by +3.6% in 2023 compared to +3.2% in death counts, indicating that the mortality burden grew faster than population changes would suggest.

### Comparison with Population Trends:

Since death rates increased while population declined (2021–2023), the data demonstrate a **genuine increase in mortality risk** rather than demographic artifacts. This reinforces the significance of the observed trend and suggests underlying factors beyond population dynamics.

### Implications:

The convergence of death rates toward 2018 levels despite population contraction warrants further investigation into potential contributing factors such as aging demographics, changes in cancer incidence, or shifts in healthcare access and reporting practices.

In [6]:
# Correlation analysis between all variables
print("=== Correlation Analysis ===")
df_sorted = df.sort_values('Year')

# Calculate correlation matrix
correlation_matrix = df_sorted[['Year', 'Deaths', 'Population', 'Age-Adjusted Rate', 'Crude Rate']].corr()

print("Correlation Matrix:")
print(correlation_matrix.round(3))

print("\nStrongest Correlations:")
# Find strongest correlations (excluding self-correlations)
correlations = []
for i in range(len(correlation_matrix.columns)):
    for j in range(len(correlation_matrix.columns)):
        if i != j:
            corr_value = correlation_matrix.iloc[i, j]
            correlations.append({
                'Variable1': correlation_matrix.columns[i],
                'Variable2': correlation_matrix.columns[j],
                'Correlation': corr_value
            })

# Sort by absolute correlation value
correlations_sorted = sorted(correlations, key=lambda x: abs(x['Correlation']), reverse=True)

# Display top correlations
for i, corr in enumerate(correlations_sorted[:5]):
    direction = "positive" if corr['Correlation'] > 0 else "negative"
    print(f"{i+1}. {corr['Variable1']} ↔ {corr['Variable2']}: {corr['Correlation']:.3f} ({direction})")

print("\nKey Insights:")
print(f"• Deaths and Population: {correlation_matrix.loc['Deaths', 'Population']:.3f}")
print(f"• Deaths and Age-Adjusted Rate: {correlation_matrix.loc['Deaths', 'Age-Adjusted Rate']:.3f}")
print(f"• Population and Age-Adjusted Rate: {correlation_matrix.loc['Population', 'Age-Adjusted Rate']:.3f}")

=== Correlation Analysis ===
Correlation Matrix:
                    Year  Deaths  Population  Age-Adjusted Rate  Crude Rate
Year               1.000   0.141      -0.219              0.071       0.293
Deaths             0.141   1.000      -0.514              0.880       0.921
Population        -0.219  -0.514       1.000             -0.781      -0.516
Age-Adjusted Rate  0.071   0.880      -0.781              1.000       0.728
Crude Rate         0.293   0.921      -0.516              0.728       1.000

Strongest Correlations:
1. Deaths ↔ Crude Rate: 0.921 (positive)
2. Crude Rate ↔ Deaths: 0.921 (positive)
3. Deaths ↔ Age-Adjusted Rate: 0.880 (positive)
4. Age-Adjusted Rate ↔ Deaths: 0.880 (positive)
5. Population ↔ Age-Adjusted Rate: -0.781 (negative)

Key Insights:
• Deaths and Population: -0.514
• Deaths and Age-Adjusted Rate: 0.880
• Population and Age-Adjusted Rate: -0.781


## Correlation Insights

Deaths are strongly positively correlated with both crude rate (0.92) and age-adjusted rate (0.88), which is expected given that mortality rates are derived from total deaths relative to population.

Population shows a moderate negative correlation with deaths (−0.51) and a strong negative correlation with age-adjusted rate (−0.78). This suggests that the recent increase in mortality occurred during a period of stable or declining population, reinforcing the interpretation that mortality burden increased independently of population growth.

Because the dataset spans only six years, these correlations should be interpreted descriptively rather than inferentially.

In [ ]:
# Age-Adjusted vs Crude Rate comparison
print("=== Age-Adjusted vs Crude Rate Comparison ===")
df_sorted = df.sort_values('Year')

# Calculate differences between rates
df_sorted['Rate_Difference'] = df_sorted['Age-Adjusted Rate'] - df_sorted['Crude Rate']
df_sorted['Rate_Percentage_Difference'] = ((df_sorted['Age-Adjusted Rate'] - df_sorted['Crude Rate']) / df_sorted['Crude Rate']) * 100

print("Rate Comparison by Year:")
for _, row in df_sorted.iterrows():
    print(f"{row['Year']}:")
    print(f"  Age-Adjusted: {row['Age-Adjusted Rate']:.1f}")
    print(f"  Crude Rate: {row['Crude Rate']:.1f}")
    print(f"  Difference: {row['Rate_Difference']:+.1f} ({row['Rate_Percentage_Difference']:+.1f}%)")

print("\nRate Analysis:")
print(f"• Average Age-Adjusted Rate: {df_sorted['Age-Adjusted Rate'].mean():.2f}")
print(f"• Average Crude Rate: {df_sorted['Crude Rate'].mean():.2f}")
print(f"• Average Difference: {df_sorted['Rate_Difference'].mean():.3f}")

print("\n=== Demographic vs Risk Assessment ===")
print("• Age-Adjusted Rate controls for age structure")
print("• Crude Rate reflects both age structure + risk")
print("• Difference indicates demographic impact")

print("\nRisk vs Demographic Decomposition:")
for i in range(1, len(df_sorted)):
    prev_year = df_sorted.iloc[i-1]
    curr_year = df_sorted.iloc[i]
    
    crude_change = ((curr_year['Crude Rate'] - prev_year['Crude Rate']) / prev_year['Crude Rate']) * 100
    age_adj_change = ((curr_year['Age-Adjusted Rate'] - prev_year['Age-Adjusted Rate']) / prev_year['Age-Adjusted Rate']) * 100
    demographic_impact = crude_change - age_adj_change
    risk_impact = age_adj_change
    
    print(f"{prev_year['Year']} → {curr_year['Year']}:")
    print(f"  Demographic Impact: {demographic_impact:+.1f}%")
    print(f"  True Risk Impact: {risk_impact:+.1f}%")

print("\nInterpretation:")
print("• If Age-Adjusted Rate > Crude Rate: Population aging effect")
print("• If Age-Adjusted Rate ≈ Crude Rate: True risk changes")
print("• Difference magnitude = Demographic impact strength")

print("\nCorrelation with Deaths:")
print(f"• Age-Adjusted Rate ↔ Deaths: {df_sorted['Age-Adjusted Rate'].corr(df_sorted['Deaths']):.3f}")
print(f"• Crude Rate ↔ Deaths: {df_sorted['Crude Rate'].corr(df_sorted['Deaths']):.3f}")

print("\nCorrelation with Population:")
print(f"• Age-Adjusted Rate ↔ Population: {df_sorted['Age-Adjusted Rate'].corr(df_sorted['Population']):.3f}")
print(f"• Crude Rate ↔ Population: {df_sorted['Crude Rate'].corr(df_sorted['Population']):.3f}")